src/misc/breeding.ts

```diff
@@ -0,0 +1,152 @@
+        const data = block.blockData;
+        if (data instanceof Waterlogged) data.setWaterlogged(false);
+        block.blockData = data;
+        playFeedingEffects(block, Material.WHEAT_SEEDS);
+        return true;
+      }
+    } else if (entity instanceof Cow) {
+      if (block.type === Material.HAY_BLOCK) {
+        block.type = Material.AIR;
+        playFeedingEffects(block, Material.HAY_BLOCK);
+        return true;
+      }
+    } else if (entity instanceof Pig) {
+      if (block.type === Material.COMPOSTER) {
+        const composterData = block.blockData as Levelled;
+        if (composterData.level >= BREED_COMPOST_LEVEL) {
+          composterData.level = composterData.level - BREED_COMPOST_LEVEL;
+          block.blockData = composterData;
+          playFeedingEffects(block, Material.COMPOSTER);
+          return true;
+        }
+      }
+    }
+  }
+
+  return false;
+}
+
+function hasWaterSource(entity: LivingEntity): boolean {
+  const blocksNearEntity = blocksInRadiusOfEntity(entity, 2);
+  for (const block of blocksNearEntity) {
+    if (block.type === Material.CAULDRON) {
+      const cauldronData = block.blockData as Levelled;
+      if (cauldronData.level >= BREED_WATER_LEVEL) {
+        cauldronData.level = cauldronData.level - BREED_WATER_LEVEL;
+        block.blockData = cauldronData;
+        playDrinkingEffects(block);
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+function blocksInRadiusOfEntity(entity: LivingEntity, radius: number): Block[] {
+  const blocks: Block[] = [];
+  const loc = entity.location;
+
+  for (let x = -radius; x <= radius; x++) {
+    for (let y = -radius; y <= radius; y++) {
+      for (let z = -radius; z <= radius; z++) {
+        blocks.push(loc.block.getRelative(x, y, z));
+      }
+    }
+  }
+
+  return blocks;
+}
+
+async function playFeedingEffects(block: Block, material: Material) {
+  const location = block.location.add(0.5, 0.8, 0.5);
+  const data = new ItemStack(material, 1);
+
+  for (let i = 0; i < 2; i++) {
+    location.world.playSound(location, Sound.ENTITY_GENERIC_EAT, 0.5, 1);
+    location.world.spawnParticle(
+      Particle.ITEM_CRACK,
+      location,
+      10,
+      0,
+      0,
+      0,
+      0.1,
+      data,
+    );
+
+    await wait(200, 'millis');
+  }
+}
+
+function playDrinkingEffects(block: Block) {
+  const location = block.location.add(0.5, 0.8, 0.5);
+
+  location.world.playSound(location, Sound.ITEM_BUCKET_EMPTY, 0.5, 1);
+  location.world.spawnParticle(
+    Particle.WATER_SPLASH,
+    location,
+    10,
+    0,
+    0,
+    0,
+    0.1,
+  );
+}
+
+registerEvent(EntityBreedEvent, (event) => {
+  if (!ANIMALS_WITH_CUSTOM_BREEDING.includes(event.entityType)) return;
+
+  const baby = event.entity as LivingEntity;
+  const mother = event.mother as LivingEntity;
+
+  const hasFood = hasFoodSource(mother);
+  const hasWater = hasWaterSource(mother);
+
+  if (!hasFood || !hasWater) baby.remove();
```

Comment:
If only food or only water is available, it will be consumed when breeding fails. Is this intentional?